File: web/src/lib/espn.ts
import "server-only";

/**
 * ESPN's public-but-undocumented API. Used for live game state only — anything
 * historical comes from the nflverse parquet store, which is authoritative.
 *
 * There is no SLA. Every call fails soft: a null return renders the page
 * without the live strip rather than erroring.
 */

const SITE = "https://site.api.espn.com/apis/site/v2/sports/football/nfl";

/**
 * Identify honestly, but without a `name/version` token.
 *
 * ESPN's edge 403s an unrecognised agent that carries one, so
 * "hashmark-analytics/0.1" was refused on every endpoint while the same string
 * minus the version is served. Verified deterministic. Exported because
 * `live.ts` calls the same API directly and drifted out of sync once already —
 * every ESPN request in the app must use this constant.
 */
export const ESPN_USER_AGENT = "hashmark-analytics (personal project)";

export type LiveTeam = {
  abbr: string | null;
  name: string | null;
  logo: string | null;
  score: number | null;
  record: string | null;
  winner: boolean;
};

export type LiveGame = {
  id: string;
  date: string;
  state: "pre" | "in" | "post";
  detail: string;
  venue: string | null;
  broadcast: string | null;
  home: LiveTeam;
  away: LiveTeam;
  downDistance: string | null;
  possession: string | null;
  lastPlay: string | null;
};

type EspnJson = Record<string, unknown>;

async function get(pathname: string, params: Record<string, string> = {}, revalidate = 30) {
  const url = new URL(`${SITE}${pathname}`);
  for (const [k, v] of Object.entries(params)) url.searchParams.set(k, v);
  try {
    const res = await fetch(url, {
      headers: { "User-Agent": ESPN_USER_AGENT },
      next: { revalidate },
    });
    if (!res.ok) return null;
    return (await res.json()) as EspnJson;
  } catch {
    return null;
  }
}

function team(competitor: EspnJson | undefined): LiveTeam {
  const t = (competitor?.team ?? {}) as EspnJson;
  const scoreRaw = competitor?.score;
  const score = Number(scoreRaw);
  const records = (competitor?.records ?? []) as EspnJson[];
  return {
    abbr: (t.abbreviation as string) ?? null,
    name: (t.displayName as string) ?? null,
    logo: (t.logo as string) ?? null,
    score: Number.isFinite(score) && scoreRaw !== undefined ? score : null,
    record: (records.find((r) => r.type === "total")?.summary as string) ?? null,
    winner: Boolean(competitor?.winner),
  };
}

export async function getScoreboard(params?: {
  week?: number;
  season?: number;
  seasonType?: number;
  dates?: string;
}): Promise<LiveGame[]> {
  const q: Record<string, string> = { limit: "100" };
  if (params?.week) q.week = String(params.week);
  if (params?.season) q.year = String(params.season);
  if (params?.seasonType) q.seasontype = String(params.seasonType);
  if (params?.dates) q.dates = params.dates;

  const data = await get("/scoreboard", q);
  const events = (data?.events ?? []) as EspnJson[];

  return events.map((ev): LiveGame => {
    const comp = ((ev.competitions ?? []) as EspnJson[])[0] ?? {};
    const competitors = (comp.competitors ?? []) as EspnJson[];
    const status = (ev.status ?? {}) as EspnJson;
    const statusType = (status.type ?? {}) as EspnJson;
    const situation = (comp.situation ?? {}) as EspnJson;
    const broadcasts = (comp.broadcasts ?? []) as EspnJson[];

    return {
      id: String(ev.id),
      date: String(ev.date),
      state: (statusType.state as LiveGame["state"]) ?? "pre",
      detail: (statusType.shortDetail as string) ?? "",
      venue: ((comp.venue as EspnJson)?.fullName as string) ?? null,
      broadcast: ((broadcasts[0]?.names as string[]) ?? [])[0] ?? null,
      home: team(competitors.find((c) => c.homeAway === "home")),
      away: team(competitors.find((c) => c.homeAway === "away")),
      downDistance: (situation.downDistanceText as string) ?? null,
      possession: (situation.possession as string) ?? null,
      lastPlay: ((situation.lastPlay as EspnJson)?.text as string) ?? null,
    };
  });
}

/** True when any game is actually in progress — drives the live badge. */
export function hasLiveGames(games: LiveGame[]): boolean {
  return games.some((g) => g.state === "in");
}

export type NewsItem = { headline: string; description: string; link: string; published: string };

export async function getNews(limit = 10): Promise<NewsItem[]> {
  const data = await get("/news", { limit: String(limit) }, 600);
  const articles = (data?.articles ?? []) as EspnJson[];
  return articles.map((a) => ({
    headline: (a.headline as string) ?? "",
    description: (a.description as string) ?? "",
    link: (((a.links as EspnJson)?.web as EspnJson)?.href as string) ?? "",
    published: (a.published as string) ?? "",
  }));
}

/** Full game detail: drives, scoring plays, box score, ESPN's own win probability. */
export async function getGameSummary(eventId: string) {
  return get("/summary", { event: eventId }, 30);
}

export type Highlight = {
  id: string;
  headline: string;
  description: string;
  duration: number | null;
  thumbnail: string | null;
  link: string;
  restricted: string[] | null;
};

/** mm:ss for a clip length. */
export function clipLength(seconds: number | null): string {
  if (!seconds || seconds < 0) return "";
  return `${Math.floor(seconds / 60)}:${String(Math.round(seconds % 60)).padStart(2, "0")}`;
}

/**
 * Highlight clips ESPN has attached to a game.
 *
 * These are cards that link out to ESPN's player, not embedded video. The
 * payload does hand over a direct CDN mp4, but playing it here would hotlink
 * their bandwidth, skip the ad the clip is monetised by, and — since a clip
 * can be licensed to a whitelist of countries — serve it to territories it was
 * never cleared for. The link is the honest use of the feed.
 *
 * Availability is narrow and worth stating plainly: clips ride along with a
 * game for days, not years. Sampling fourteen 2025 regular-season games six
 * months on returned zero; games from the previous night returned four or five
 * each. So this renders during and just after a game, and is absent from the
 * archive by design rather than by failure.
 */
export async function getGameHighlights(eventId: string): Promise<Highlight[]> {
  const data = await get("/summary", { event: eventId }, 300);
  const videos = (data?.videos ?? []) as EspnJson[];
  const now = Date.now();

  return videos
    .map((v): Highlight | null => {
      const links = (v.links ?? {}) as EspnJson;
      const link = ((links.web as EspnJson)?.href as string) ?? "";
      if (!link) return null;

      // Respect the feed's own windows rather than showing a dead card.
      const time = (v.timeRestrictions ?? {}) as EspnJson;
      const embargo = time.embargoDate ? Date.parse(String(time.embargoDate)) : null;
      const expires = time.expirationDate ? Date.parse(String(time.expirationDate)) : null;
      if (embargo && Number.isFinite(embargo) && now < embargo) return null;
      if (expires && Number.isFinite(expires) && now > expires) return null;

      const geo = (v.geoRestrictions ?? {}) as EspnJson;
      const countries = (geo.countries as string[]) ?? [];

      return {
        id: String(v.id ?? link),
        headline: (v.headline as string) ?? "",
        description: (v.description as string) ?? "",
        duration: Number.isFinite(Number(v.duration)) ? Number(v.duration) : null,
        thumbnail: (v.thumbnail as string) ?? null,
        link,
        restricted: geo.type === "whitelist" && countries.length ? countries : null,
      };
    })
    .filter((v): v is Highlight => v !== null);
}

File: pipeline/nflx/sources/espn.py
"""ESPN's undocumented public API.

Used for things nflverse does not carry: live in-game state, drive detail,
news, venue/broadcast metadata and team logos at multiple resolutions.

There is no SLA here. Every call is wrapped so a failure degrades to None
rather than taking a build down, and responses are cached on disk.
"""

from __future__ import annotations

import json
import time
from pathlib import Path
from typing import Any

import requests

from ..config import CACHE

SITE = "https://site.api.espn.com/apis/site/v2/sports/football/nfl"
CORE = "https://sports.core.api.espn.com/v2/sports/football/leagues/nfl"
CDN = "https://cdn.espn.com/core/nfl"
FANTASY = "https://lm-api-reads.fantasy.espn.com/apis/v3/games/ffl/seasons"

# Identify honestly, but without a `name/version` token: ESPN's edge 403s any
# unrecognised agent carrying one, so "hashmark-analytics/0.1 (…)" was being
# refused on every endpoint while the same string minus a version is served.
# Verified deterministic — 4/4 refusals with the version, 4/4 responses without.
HEADERS = {"User-Agent": "gridiron-analytics (personal project)"}
TIMEOUT = 20


def _get(url: str, params: dict | None = None, cache_key: str | None = None,
         ttl: int = 0) -> Any | None:
    """GET JSON with optional on-disk caching. Returns None on any failure."""
    cache_path: Path | None = None
    if cache_key:
        cache_path = CACHE / f"espn_{cache_key}.json"
        if cache_path.exists() and (ttl == 0 or time.time() - cache_path.stat().st_mtime < ttl):
            try:
                return json.loads(cache_path.read_text(encoding="utf-8"))
            except json.JSONDecodeError:
                pass
    try:
        r = requests.get(url, params=params, headers=HEADERS, timeout=TIMEOUT)
        r.raise_for_status()
        data = r.json()
    except (requests.RequestException, json.JSONDecodeError) as exc:
        print(f"  ! espn {url} failed: {exc}", flush=True)
        return None
    if cache_path is not None:
        cache_path.write_text(json.dumps(data), encoding="utf-8")
    return data


# ----------------------------------------------------------------- endpoints

def scoreboard(dates: str | None = None, week: int | None = None,
               season: int | None = None, season_type: int = 2) -> Any | None:
    """Live/completed games. `dates` is YYYYMMDD or YYYYMMDD-YYYYMMDD."""
    params: dict[str, Any] = {"limit": 100}
    if dates:
        params["dates"] = dates
    if week:
        params["week"] = week
    if season:
        params["seasontype"] = season_type
        params["year"] = season
    return _get(f"{SITE}/scoreboard", params)


def summary(event_id: str) -> Any | None:
    """Full game detail: drives, plays, box score, leaders, win probability."""
    return _get(f"{SITE}/summary", {"event": event_id})


def standings(season: int) -> Any | None:
    return _get(
        f"{CDN}/standings",
        {"xhr": 1, "season": season},
        cache_key=f"standings_{season}",
        ttl=3600,
    )


def teams_list() -> Any | None:
    return _get(f"{SITE}/teams", cache_key="teams", ttl=86400)


def team_detail(team_id: str | int) -> Any | None:
    return _get(f"{SITE}/teams/{team_id}", cache_key=f"team_{team_id}", ttl=86400)


def team_roster(team_id: str | int) -> Any | None:
    return _get(f"{SITE}/teams/{team_id}/roster", cache_key=f"roster_{team_id}", ttl=86400)


def athlete(espn_id: str | int) -> Any | None:
    """Player bio, headshot and current team."""
    return _get(
        f"{CORE}/athletes/{espn_id}",
        cache_key=f"athlete_{espn_id}",
        ttl=86400 * 7,
    )


def news(limit: int = 30) -> Any | None:
    return _get(f"{SITE}/news", {"limit": limit})


# ----------------------------------------------------------------- shaping

def normalize_scoreboard(payload: Any) -> list[dict]:
    """Flatten an ESPN scoreboard into the shape the app renders."""
    if not payload:
        return []
    out: list[dict] = []
    for ev in payload.get("events", []):
        comp = (ev.get("competitions") or [{}])[0]
        status = (ev.get("status") or {}).get("type", {})
        teams: dict[str, dict] = {}
        for c in comp.get("competitors", []):
            side = c.get("homeAway", "home")
            team = c.get("team", {})
            teams[side] = {
                "abbr": team.get("abbreviation"),
                "name": team.get("displayName"),
                "logo": team.get("logo"),
                "color": f"#{team['color']}" if team.get("color") else None,
                "score": int(c["score"]) if str(c.get("score", "")).isdigit() else None,
                "record": next(
                    (r.get("summary") for r in c.get("records", []) if r.get("type") == "total"),
                    None,
                ),
                "winner": c.get("winner"),
            }
        situation = comp.get("situation") or {}
        out.append({
            "id": ev.get("id"),
            "date": ev.get("date"),
            "week": (payload.get("week") or {}).get("number"),
            "state": status.get("state"),            # pre | in | post
            "detail": status.get("shortDetail"),
            "period": (ev.get("status") or {}).get("period"),
            "clock": (ev.get("status") or {}).get("displayClock"),
            "venue": (comp.get("venue") or {}).get("fullName"),
            "broadcast": next(
                (b.get("names", [None])[0] for b in comp.get("broadcasts", [])), None
            ),
            "home": teams.get("home"),
            "away": teams.get("away"),
            "down_distance": situation.get("downDistanceText"),
            "possession": situation.get("possession"),
            "last_play": (situation.get("lastPlay") or {}).get("text"),
        })
    return out


# ESPN's own position ids, which do not match anybody else's.
FANTASY_POSITIONS = {1: "QB", 2: "RB", 3: "WR", 4: "TE", 5: "K", 16: "DST"}


def fantasy_players(season: int) -> list[dict] | None:
    """ESPN's own draft ranks and league-wide ADP for a fantasy season.

    This is what an ESPN league shows its drafters by default, which is the
    reason it is worth having separately from a consensus ranking: it is not
    another opinion, it is the behaviour of the room you are drafting against.

    The endpoint ignores the limit in the filter and returns every player, so
    the caller filters. Ranks are absent for players ESPN does not rank at all.
    """
    filt = json.dumps(
        {"players": {"limit": 5000,
                     "sortDraftRanks": {"sortPriority": 1, "sortAsc": True, "value": "PPR"}}}
    )
    url = f"{FANTASY}/{season}/players?view=kona_player_info"
    try:
        r = requests.get(
            url,
            headers={**HEADERS, "x-fantasy-filter": filt},
            timeout=90,
        )
        r.raise_for_status()
        payload = r.json()
    except (requests.RequestException, json.JSONDecodeError) as exc:
        print(f"  espn fantasy ranks unavailable: {exc!r}")
        return None
    return payload if isinstance(payload, list) else payload.get("players")
